GitError Class:Zlib on cloning General

I've been fairly frequently getting a mysterious `GitError(Code:ERROR, Class:Zlib, error inflating zlib stream)` on julia-nightly while using julia-buildpkg.

I think this might possibly only occur on Juila-nightly. Example:

https://github.com/JuliaArrays/StaticArrays.jl/pull/840/checks?check_run_id=1272952752

```
Run julia-actions/julia-buildpkg@latest
     Cloning registry from "https://github.com/JuliaRegistries/General.git"
25l    Fetching: [>                                        ]  0.0 %
    Fetching: [==========>                              ]  22.5 %
    Fetching: [================>                        ]  38.2 %
    Fetching: [=========================>               ]  60.5 %
ERROR: failed to clone from https://github.com/JuliaRegistries/General.git, error: GitError(Code:ERROR, Class:Zlib, error inflating zlib stream)
Stacktrace:
  [1] pkgerror(msg::String)
    @ Pkg.Types /buildworker/worker/package_linux32/build/usr/share/julia/stdlib/v1.6/Pkg/src/Types.jl:52
  [2] clone(ctx::Pkg.Types.Context, url::String, source_path::String; header::String, credentials::Nothing, kwargs::Base.Iterators.Pairs{Union{}, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
    @ Pkg.GitTools /buildworker/worker/package_linux32/build/usr/share/julia/stdlib/v1.6/Pkg/src/GitTools.jl:153
  [3] (::Pkg.Types.var"#82#85"{Pkg.Types.Context, String, Pkg.Types.RegistrySpec})(tmp::String)
    @ Pkg.Types /buildworker/worker/package_linux32/build/usr/share/julia/stdlib/v1.6/Pkg/src/Types.jl:998
  [4] mktempdir(fn::Pkg.Types.var"#82#85"{Pkg.Types.Context, String, Pkg.Types.RegistrySpec}, parent::String; prefix::String)
    @ Base.Filesystem ./file.jl:711
  [5] mktempdir
    @ ./file.jl:709 [inlined]
  [6] clone_or_cp_registries(ctx::Pkg.Types.Context, regs::Vector{Pkg.Types.RegistrySpec}, depot::String)
    @ Pkg.Types /buildworker/worker/package_linux32/build/usr/share/julia/stdlib/v1.6/Pkg/src/Types.jl:978
  [7] clone_or_cp_registries
    @ /buildworker/worker/package_linux32/build/usr/share/julia/stdlib/v1.6/Pkg/src/Types.jl:971 [inlined]
  [8] #add#6
    @ /buildworker/worker/package_linux32/build/usr/share/julia/stdlib/v1.6/Pkg/src/Registry.jl:32 [inlined]
  [9] add
    @ /buildworker/worker/package_linux32/build/usr/share/julia/stdlib/v1.6/Pkg/src/Registry.jl:28 [inlined]
 [10] #add#5
    @ /buildworker/worker/package_linux32/build/usr/share/julia/stdlib/v1.6/Pkg/src/Registry.jl:26 [inlined]
 [11] add
    @ /buildworker/worker/package_linux32/build/usr/share/julia/stdlib/v1.6/Pkg/src/Registry.jl:26 [inlined]
 [12] add(regs::Vector{String}; kwargs::Base.Iterators.Pairs{Union{}, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
    @ Pkg.Registry /buildworker/worker/package_linux32/build/usr/share/julia/stdlib/v1.6/Pkg/src/Registry.jl:25
 [13] add
    @ /buildworker/worker/package_linux32/build/usr/share/julia/stdlib/v1.6/Pkg/src/Registry.jl:25 [inlined]
 [14] #add#1
    @ /buildworker/worker/package_linux32/build/usr/share/julia/stdlib/v1.6/Pkg/src/Registry.jl:24 [inlined]
 [15] add(reg::String)
    @ Pkg.Registry /buildworker/worker/package_linux32/build/usr/share/julia/stdlib/v1.6/Pkg/src/Registry.jl:24
 [16] top-level scope
    @ none:1

caused by: GitError(Code:ERROR, Class:Zlib, error inflating zlib stream)
Stacktrace:
  [1] macro expansion
    @ /buildworker/worker/package_linux32/build/usr/share/julia/stdlib/v1.6/LibGit2/src/error.jl:110 [inlined]
  [2] clone(repo_url::SubString{String}, repo_path::String, clone_opts::LibGit2.CloneOptions)
    @ LibGit2 /buildworker/worker/package_linux32/build/usr/share/julia/stdlib/v1.6/LibGit2/src/repository.jl:459
  [3] clone(repo_url::SubString{String}, repo_path::String; branch::String, isbare::Bool, remote_cb::Ptr{Nothing}, credentials::LibGit2.CachedCredentials, callbacks::Dict{Symbol, Tuple{Ptr{Nothing}, Any}})
    @ LibGit2 /buildworker/worker/package_linux32/build/usr/share/julia/stdlib/v1.6/LibGit2/src/LibGit2.jl:580
  [4] clone(ctx::Pkg.Types.Context, url::String, source_path::String; header::String, credentials::Nothing, kwargs::Base.Iterators.Pairs{Union{}, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
    @ Pkg.GitTools /buildworker/worker/package_linux32/build/usr/share/julia/stdlib/v1.6/Pkg/src/GitTools.jl:143
  [5] (::Pkg.Types.var"#82#85"{Pkg.Types.Context, String, Pkg.Types.RegistrySpec})(tmp::String)
    @ Pkg.Types /buildworker/worker/package_linux32/build/usr/share/julia/stdlib/v1.6/Pkg/src/Types.jl:998
  [6] mktempdir(fn::Pkg.Types.var"#82#85"{Pkg.Types.Context, String, Pkg.Types.RegistrySpec}, parent::String; prefix::String)
    @ Base.Filesystem ./file.jl:711
  [7] mktempdir
    @ ./file.jl:709 [inlined]
  [8] clone_or_cp_registries(ctx::Pkg.Types.Context, regs::Vector{Pkg.Types.RegistrySpec}, depot::String)
    @ Pkg.Types /buildworker/worker/package_linux32/build/usr/share/julia/stdlib/v1.6/Pkg/src/Types.jl:978
  [9] clone_or_cp_registries
    @ /buildworker/worker/package_linux32/build/usr/share/julia/stdlib/v1.6/Pkg/src/Types.jl:971 [inlined]
 [10] #add#6
    @ /buildworker/worker/package_linux32/build/usr/share/julia/stdlib/v1.6/Pkg/src/Registry.jl:32 [inlined]
 [11] add
    @ /buildworker/worker/package_linux32/build/usr/share/julia/stdlib/v1.6/Pkg/src/Registry.jl:28 [inlined]
 [12] #add#5
    @ /buildworker/worker/package_linux32/build/usr/share/julia/stdlib/v1.6/Pkg/src/Registry.jl:26 [inlined]
 [13] add
    @ /buildworker/worker/package_linux32/build/usr/share/julia/stdlib/v1.6/Pkg/src/Registry.jl:26 [inlined]
 [14] add(regs::Vector{String}; kwargs::Base.Iterators.Pairs{Union{}, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
    @ Pkg.Registry /buildworker/worker/package_linux32/build/usr/share/julia/stdlib/v1.6/Pkg/src/Registry.jl:25
 [15] add
    @ /buildworker/worker/package_linux32/build/usr/share/julia/stdlib/v1.6/Pkg/src/Registry.jl:25 [inlined]
 [16] #add#1
    @ /buildworker/worker/package_linux32/build/usr/share/julia/stdlib/v1.6/Pkg/src/Registry.jl:24 [inlined]
 [17] add(reg::String)
    @ Pkg.Registry /buildworker/worker/package_linux32/build/usr/share/julia/stdlib/v1.6/Pkg/src/Registry.jl:24
 [18] top-level scope
    @ none:1
25h
Error: Process completed with exit code 1.
```

--- a/action.yml
+++ b/action.yml
@@ -17,7 +17,7 @@ runs:
     # * https://github.com/JuliaLang/Pkg.jl/issues/2011
     # * https://github.com/JuliaRegistries/General/issues/16777
     # * https://github.com/JuliaPackaging/PkgServer.jl/issues/60
-  - run: julia --color=yes -e 'using Pkg; VERSION >= v"1.5-" && !isdir(joinpath(DEPOT_PATH[1], "registries", "General")) && Pkg.Registry.add("General")'
+  - run: julia --color=yes "$GITHUB_ACTION_PATH"/add_general_registry.jl
     shell: bash
     env:
       # We set `JULIA_PKG_SERVER` only for this step to enforce
@@ -25,5 +25,6 @@ runs:
       # the request metadata to pkg.julialang.org when installing
       # packages via `Pkg.test`.
       JULIA_PKG_SERVER: ""
+
   - run: julia --color=yes --project -e 'using Pkg; if VERSION >= v"1.1.0-rc1"; Pkg.build(verbose=true); else Pkg.build(); end'
     shell: bash

--- a/add_general_registry.jl
+++ b/add_general_registry.jl
@@ -1,0 +1,47 @@
+using Pkg
+
+function general_registry_location()
+    general_registry_dir = joinpath(DEPOT_PATH[1], "registries", "General")
+    registry_toml_file = joinpath(general_registry_dir, "Registry.toml")
+    return general_registry_dir, registry_toml_file
+end
+
+function general_registry_exists()
+    general_registry_dir, registry_toml_file = general_registry_location()
+    if !isdir(general_registry_dir)
+        return false
+    elseif !isfile(registry_toml_file)
+        return false
+    else
+        return true
+    end
+end
+
+function add_general_registry()
+    @info("Attempting to clone the General registry")
+    general_registry_dir, registry_toml_file = general_registry_location()
+    rm(general_registry_dir; force = true, recursive = true)
+    Pkg.Registry.add("General")
+    isfile(registry_toml_file) || throw(ErrorException("the Registry.toml file does not exist"))
+    return nothing
+end
+
+function main(; n = 10, max_delay = 120)
+    if VERSION >= v"1.5-"
+        if !general_registry_exists()
+            delays = ExponentialBackOff(; n = n, max_delay = max_delay)
+            try
+                retry(add_general_registry; delays = delays)()
+                @info("Successfully added the General registry")
+            catch ex
+                msg = "I was unable to added the General registry. However, the build will continue."
+                @error(msg, exception=(ex,catch_backtrace()))
+            end
+        else
+            @info("The General registry already exists locally")
+        end
+    end
+    return nothing
+end
+
+main()

--- a/add_general_registry.jl
+++ b/add_general_registry.jl
@@ -39,7 +39,7 @@ function main(; n = 10, max_delay = 120)
         retry(add_general_registry; delays = delays)()
         @info("Successfully added the General registry")
     catch ex
-        msg = "I was unable to added the General registry. However, the build will continue."
+        msg = "I was unable to add the General registry. However, the build will continue."
         @error(msg, exception=(ex,catch_backtrace()))
     end
 

--- a/add_general_registry.jl
+++ b/add_general_registry.jl
@@ -27,21 +27,23 @@ function add_general_registry()
 end
 
 function main(; n = 10, max_delay = 120)
-    if VERSION >= v"1.5-"
-        if !general_registry_exists()
-            delays = ExponentialBackOff(; n = n, max_delay = max_delay)
-            try
-                retry(add_general_registry; delays = delays)()
-                @info("Successfully added the General registry")
-            catch ex
-                msg = "I was unable to added the General registry. However, the build will continue."
-                @error(msg, exception=(ex,catch_backtrace()))
-            end
-        else
-            @info("The General registry already exists locally")
-        end
+    VERSION >= v"1.5-" || return
+
+    if general_registry_exists()
+        @info("The General registry already exists locally")
+        return
     end
-    return nothing
+
+    delays = ExponentialBackOff(; n = n, max_delay = max_delay)
+    try
+        retry(add_general_registry; delays = delays)()
+        @info("Successfully added the General registry")
+    catch ex
+        msg = "I was unable to added the General registry. However, the build will continue."
+        @error(msg, exception=(ex,catch_backtrace()))
+    end
+
+    return
 end
 
 main()

